ShaiRandom.UnitTests/GapShufflerTests.cs

```diff
@@ -0,0 +1,156 @@
+            var shuffler = rng.InPlaceGapShuffler(list);
+
+            var shuffled = shuffler.Take(10).ToArray();
+            CheckResult(originalList, list, shuffled, true);
+        }
+
+        [Fact]
+        public void GapShuffleInPlaceNextIList()
+        {
+            var rng = new MizuchiRandom(1);
+
+            var originalList = new[] { 1, 2, 3, 4, 5 };
+            var list = new[] { 1, 2, 3, 4, 5 };
+            var shuffler = rng.InPlaceGapShuffler(list);
+
+            var shuffled = GetShuffledWithNext(shuffler);
+            CheckResult(originalList, list, shuffled, true);
+        }
+
+        [Fact]
+        public void GapShuffleInPlaceMemory()
+        {
+            var rng = new MizuchiRandom(1);
+
+            var originalList = new[] { 1, 2, 3, 4, 5 };
+            var list = new[] { 1, 2, 3, 4, 5 };
+            var shuffler = rng.InPlaceGapShuffler(list.AsMemory());
+
+            var shuffled = shuffler.Take(10).ToArray();
+            CheckResult(originalList, list, shuffled, true);
+        }
+
+        [Fact]
+        public void GapShuffleInPlaceNextMemory()
+        {
+            var rng = new MizuchiRandom(1);
+
+            var originalList = new[] { 1, 2, 3, 4, 5 };
+            var list = new[] { 1, 2, 3, 4, 5 };
+            var shuffler = rng.InPlaceGapShuffler(list.AsMemory());
+
+            var shuffled = GetShuffledWithNext(shuffler);
+            CheckResult(originalList, list, shuffled, true);
+        }
+
+        private static int[] GetShuffledWithNext(GapShufflerEnumerator<int> shuffler)
+        {
+            var array = new int[10];
+            for (int i = 0; i < array.Length; i++)
+                array[i] = shuffler.Next();
+
+            return array;
+        }
+
+        private static int[] GetShuffledWithNext(GapShufflerInPlaceMemoryEnumerator<int> shuffler)
+        {
+            var array = new int[10];
+            for (int i = 0; i < array.Length; i++)
+                array[i] = shuffler.Next();
+
+            return array;
+        }
+
+        private static void CheckResult(int[] originalList, int[] currentList, int[] enumerableToArray, bool inPlace)
+        {
+            // Only required for how we test, not for the algorithm itself
+            Assert.Equal(originalList.Length * 2, enumerableToArray.Length);
+
+            // Split into two halves, each of which should contain each element of the original list exactly once
+            var halves = new[] { enumerableToArray[..5], enumerableToArray[5..] };
+            foreach (var half in halves)
+            {
+                // Each set of 5 contains each element of the original precisely once
+                Assert.Equal(originalList.ToHashSet(), half.ToHashSet());
+
+                // No set is the original list.  This is not _technically_ enforced by the algorithm, but it's
+                // highly unlikely to happen by chance (and won't happen with the particular RNG seed we're using).
```

Comment:
This is definitely a little hacky, but since MizuchiRandom should be stable and the code to shuffle or get bounded ints should be too, relying on the seed being one of the vast majority of seeds that have this behavior should be fine.